xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -259,35 +258,87 @@ public void run() {
         }
         ResourceSubscriber<T> subscriber =
             (ResourceSubscriber<T>) resourceSubscribers.get(type).get(resourceName);
+
         if (subscriber == null) {
           logger.log(XdsLogLevel.INFO, "Subscribe {0} resource {1}", type, resourceName);
           subscriber = new ResourceSubscriber<>(type, resourceName);
           resourceSubscribers.get(type).put(resourceName, subscriber);
-          if (subscriber.controlPlaneClient != null) {
-            subscriber.controlPlaneClient.adjustResourceSubscription(type);
+          subscriber.addWatcher(watcher, watcherExecutor);
+          boolean hadErrorBeforeCpc = subscriber.errorDescription != null;
+
+          ControlPlaneClient cpcToUse = subscriber.loadControlPlane();
+
+          if (subscriber.errorDescription != null) {
```

Comment:
This condition can leak a freshly-created `cpcToUse`. Move `loadControlPlane()` after the condition?